### ISSUE TITLE
Misbehavior (with fix) on iOS when using FrameBuffer

### ISSUE THREAD

**Issue by b3nk4n** [libgdx/libgdx#7297]
#### Issue details

I've been facing an issue on iOS only when using `FrameBuffer`. When I ported my game [October Bro](https://play.google.com/store/apps/details?id=de.bsautermeister.jump) from Android/Desktop to iOS, I noticed that my game screen was just black (or just the screen-clear color I used).

After longer investigation and digging into the LibGdx code, I could figure out the problem, and come up with a workaround. And my hope is that this issue report turns into a simple fix PR (either by me or by you), so that my workaround will not be needed anymore.

#### Reproduction steps/code

Use the following code on iOS, and toggle the `FIX` flag.

```java
public class MyGdxGame extends ApplicationAdapter {

	private static final boolean FIX = true; // <-- please toggle me! :)

	SpriteBatch batch;
	Texture img;

	FrameBuffer frameBuffer;

	OrthographicCamera camera;

	boolean created = false;

	@Override
	public void create () {
		batch = new SpriteBatch();
		img = new Texture("badlogic.jpg");

		camera = new OrthographicCamera();

		created = true;
		frameBuffer = new FrameBuffer(Pixmap.Format.RGBA8888, Gdx.graphics.getWidth(), Gdx.graphics.getHeight(), false);
	}

	@Override
	public void render () {
		Gdx.gl.glClearColor(1f, 0f, 0f, 1);
		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

		IntBuffer oldFbo = BufferUtils.newIntBuffer(1);
		Gdx.gl.glGetIntegerv(GL_FRAMEBUFFER_BINDING, oldFbo);

		frameBuffer.begin();

		Gdx.gl.glClearColor(0f, 1f, 1f, 1);
		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

		camera.setToOrtho(true, frameBuffer.getWidth(), frameBuffer.getHeight());
		camera.update();

		Gdx.gl.glClearColor(1.0f, 1.0f, 1.0f, 255 / 255.0f);

		batch.begin();
		batch.draw(img, 0, 0);
		batch.end();

		frameBuffer.end();

		if (FIX) {
			// use the previous back buffer index, instead of defaultFramebufferHandle from GLFrameBuffer#unbind()
			int fboIdx = oldFbo.get();
			Gdx.gl.glBindFramebuffer(GL_FRAMEBUFFER, fboIdx);
		}

		camera.setToOrtho(true, Gdx.graphics.getWidth(), Gdx.graphics.getHeight());
		camera.update();

		// render pixel-art buffer to screen
		batch.setProjectionMatrix(camera.combined);
		batch.begin();
		batch.draw(frameBuffer.getColorBufferTexture(),
				0, 0, 0, 0, Gdx.graphics.getWidth() / 2f, Gdx.graphics.getHeight() / 2f, 1f, 1f, 0f, 0, 0,
				frameBuffer.getWidth(), frameBuffer.getHeight(), false, true);
		batch.end();
	}

	@Override
	public void dispose () {
		img.dispose();
		batch.dispose();
		if (frameBuffer != null) {
			frameBuffer.dispose();
		}
	}
}
```

The problematic code seems to come from the following section, which seems to even try to exactly this iOS specific problem according to the comment:
https://github.com/libgdx/libgdx/blob/9036b801cc3f4844d346dc96bf9b8f89d8390ad0/gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java#L128-L138

However, while testing and debugging this, I could observe the following:
- As written in these lines, the screen is always `fboIdx = defaultFramebufferHandle = 0` on Desktop/Android, but usually `fboIdx = 1` on iOS
- However, that might depend on the game. Or on the point in time when the `new FrameBuffer(...)` constructor is called for the first time. I observed on my single iOS game that the very first time the constructor is called, that the stored fboIndex is first `0`, and later it would be `1` on iOS.
- The fact that via the static `defaultFramebufferHandleInitialized` the initialization of the `defaultFramebufferHandle` is only done once causes the problem that we **might try to go back to the wrong index** on `FrameBuffer#end()`. As shown in the above code example.

#### TL;DR Workaround

Summarizing my current workaround in a brief from above, the problem can be prevented by wrapping the `frameBuffer` begin/end function with:

```java

IntBuffer oldFbo = BufferUtils.newIntBuffer(1);
Gdx.gl.glGetIntegerv(GL_FRAMEBUFFER_BINDING, oldFbo);

frameBuffer.begin();
// ...
frameBuffer.end();

int fboIdx = oldFbo.get();
Gdx.gl.glBindFramebuffer(GL_FRAMEBUFFER, fboIdx);
```

To kind of override the internal behavior of how `FrameBuffer#end()` does the unbind to go back to the screen.

#### Version of libGDX and/or relevant dependencies
`gdxVersion = '1.12.0'`

#### Screenshots

With both `FIX=false` or `FIX=true`, there is no problem on Android/Desktop:

<img width="752" alt="Screenshot 2023-12-09 at 1 33 23 PM" src="https://github.com/libgdx/libgdx/assets/2537736/c662a250-3005-430d-b88e-0f690b816738">

However, on iOS, using `FIX=false` causes a wrong result:

![Simulator Screenshot - iPhone 15 - 2023-12-09 at 13 32 31](https://github.com/libgdx/libgdx/assets/2537736/8e6f218d-fac8-452f-8174-57bc1cea2342)

The reason is that `FrameBuffer#end()` does **not** work as expected.

With my fix to enforce going back to the previous frame buffer (here: the screen), it works correctly on **any** platform, even iOS:

![Simulator Screenshot - iPhone 15 - 2023-12-09 at 13 33 00](https://github.com/libgdx/libgdx/assets/2537736/0356d594-f55a-428d-bf69-f36610610cf3)

#### Please select the affected platforms
- [ ] Android
- [x] iOS
- [ ] HTML/GWT
- [ ] Windows
- [ ] Linux
- [ ] macOS

#### References:

- [FBO bug in Cocos2D](https://stackoverflow.com/questions/11617013/why-would-glbindframebuffergl-framebuffer-0-result-in-blank-screen-in-cocos2d)

**Comment by b3nk4n** [comment]
Personally, I'm unblocked by my workaround. However, would be happy to provide a PR for this reported problem. Before doing so, I would however like to get some questions clarified before I would create a PR, to have less back and forth as part of the final PR:

**Questions:**
1. Is there any strict need to only set the [defaultFramebufferHandle](https://github.com/libgdx/libgdx/blob/9036b801cc3f4844d346dc96bf9b8f89d8390ad0/gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java#L134) only once via [defaultFramebufferHandleInitialized](https://github.com/libgdx/libgdx/blob/9036b801cc3f4844d346dc96bf9b8f89d8390ad0/gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java#L129C8-L129C43), instead of whenever an FBO is created? Are e.g. [these OpenGL functions to get the FRAMEBUFFER_BINDING](https://github.com/libgdx/libgdx/blob/9036b801cc3f4844d346dc96bf9b8f89d8390ad0/gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java#L132-L133) expensive?
2. Would you prefer fixing this in a way that yields iOS specific special handling, as in the current implementation [here](https://github.com/libgdx/libgdx/blob/9036b801cc3f4844d346dc96bf9b8f89d8390ad0/gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java#L128-L138)? Even when the fix (or iOS special handling) also would work fine with other platforms? Or instead better have a single implementation that works for all the same way. Even when there might be a slightly more optimal implementation for specific platforms (e.g. simply assume [defaultFramebufferHandle = 0](https://github.com/libgdx/libgdx/blob/9036b801cc3f4844d346dc96bf9b8f89d8390ad0/gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java#L136C5-L136C33) if platform is not iOS)

Thank you in advance!

**Comment by Tom-Ski** [comment]
Tested this on device, and it does not reproduce. Are you testing on sim here?

**Comment by Tom-Ski** [comment]
Looked closer into this, and Ok this is nuts! 👁️ 

I think this was introduced by 0f24d72bd4c86b52531b4e4d60ccdf58c04db98c as we have changed creation callback from waiting for the first frame, but I'm still super surprised that it hasn't been reported before you.

I guess most of us are not creating framebuffers on create, (usually after some splash screen which doesn't require FBO would be my guess)

Its not apparent in the test bench, since they construct their framebuffers after a first render due to structure of test suite. This has to happen in the create of the Adapater. We need to wait for the very first frame before finding the 'default' render buffer handle it seems as it returns 0 until we get that first frame.

I'd rather not query something on every single interaction with framebuffers, every little helps with not putting gl commands into the buffers.  Perhaps we can bring back the first frame logic, even if its for setting the default handle.
Should take a look at https://github.com/libgdx/libgdx/issues/6971 again and see if we can find a better solution for all.

**Comment by b3nk4n** [comment]
> Tested this on device, and it does not reproduce. Are you testing on sim here?

I tested this on both a simulator with iOS 17.0 on an iPhone 15 with the following simulator version:

```
Version 15.0.1 (1015.2)
SimulatorKit 935.1
CoreSimulator 932.2
```

And my using iOS 17.1.1 on my physical iPhone 13 device.

Thx @Tom-Ski for all the details above 👍 
If I understand correctly, if I would skip creating the FBO in just the very first frame, then things would work correctly (**Edit:** and I can confirm that this is working that way when I create the FBO lazily in the second frame).

Sounds more complicated that I thought. And would therefore better leave that to you, as you seem to have a lot of context around this already. But let me know in case there is anything I can help with.
Thx!

**Comment by LobbyDivinus** [comment]
So this will be fixed? Nice. I also stumbled across this issue when using shadow light which internally uses a framebuffer. Postponing the creation until the first update call works but doesn't feel right.
